docs/posts/lms-for-data/index.qmd

```diff
@@ -0,0 +1,695 @@
+    `table_names`, `table_schemas`, and `table_previews`.
+
+    Use newlines and indentation for readability.
+    """  # <1
+
+
+def text_to_sql(  # <2>
+    text: str,  # <2>
+    table_names: list[str],  # <2>
+    table_schemas: list[str],  # <2>
+    table_previews: list[str],  # <2>
+) -> str:  # <2>
+    sql = _text_to_sql(text, table_names, table_schemas, table_previews)  # <3>
+    return sql.strip().strip(";")  # <4>
+```
+
+1. Define a Marvin function to write SQL from text
+2. Define a Python function to apply the Marvin function to a string
+3. Generate the SQL string
+4. Strip the trailing whitespace and semicolon that is sometimes generated, as
+ it causes issues with Ibis
+
+We can try that out on our penguins table:
+
+```{python}
+text = (  # <1>
+    "the count of penguins by species, from highest to lowest, per each island"  # <1>
+)
+
+table_names = ["penguins"]  # <2>
+table_schemas = [str(penguins.schema())]  # <2>
+table_previews = [str(penguins.limit(5))]  # <2>
+
+sql = text_to_sql(text, table_names, table_schemas, table_previews)  # <3>
+print(sql)
+```
+
+1. Create a natural language query
+2. Provide the table names, schemas, and previews
+3. Generate the SQL string
+
+And execute the SQL:
+
+```{python}
+r = penguins.sql(sql) # <1>
+r
+```
+
+1. Execute the SQL string on the table
+
+:::{.callout-note title="This is definitely RAG"}
+In this case, we've retrieved a table's name, schema, and a preview of its data
+to generate a high-quality SQL query. We're definitely doing RAG now.
+
+While most RAG implementations use a similarity metric between text converted to
+numberes (tokens/word embeddings), I find simple text retrieval in a
+well-organized knowledge base to be more effective. We'll follow up on this in
+future posts.
+:::
+
+#### A more complex example
+
+Let's see how this works on a query that requires joining two tables. We'll load
+in some IMDB data:
+
+```{python}
+imdb_title_basics = ibis.examples.imdb_title_basics.fetch()  # <1>
+imdb_title_basics
+```
+
+1. Load the IMDB title basics dataset from Ibis examples
+
+```{python}
+imdb_title_ratings = ibis.examples.imdb_title_ratings.fetch()  # <1>
+imdb_title_ratings
+```
+
+1. Load the IMDB title ratings dataset from Ibis examples
+
+```{python}
+text = "the highest rated movies w/ over 100k ratings -- movies only"  # <1>
+
+table_names = ["imdb_title_basics", "imdb_title_ratings"]  # <2>
+table_schemas = [
+    str(imdb_title_basics.schema()),
+    str(imdb_title_ratings.schema()),
+]
+table_previews = [
+    str(imdb_title_basics.limit(5)),
+    str(imdb_title_ratings.limit(5)),
+]  # <2>
+
+sql = text_to_sql(text, table_names, table_schemas, table_previews)  # <3>
+print(sql)
+```
+
+1. Create a natural language query
+2. Provide the table names, schemas, and previews
+3. Generate the SQL string
+
+```{python}
+r = imdb_title_basics.sql(sql)  # <1>
+r
+```
+
+1. Execute the SQL string on the table
+
+## Issues with language models today
```

Comment:
not sure how deep you want to go into this, but there are many other issues. like hallucinations: when you can't trust the generated responses